### PR TITLE
feat: some `Char`, `UInt`, and `Fin` theorems

### PR DESCRIPTION
**Pull request by leodemoura**
for SSFT24 summer school: https://github.com/david-christiansen/ssft24

### CODE REVIEW

File: src/Init/Data/UInt/Lemmas.lean [hunk 1]
```diff
@@ -0,0 +1,90 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Leonardo de Moura
+-/
+prelude
+import Init.Data.UInt.Basic
+import Init.Data.Fin.Lemmas
+
+-- Should we create a macro?
```
Comment (digama0) [inline src/Init/Data/UInt/Lemmas.lean:10]:
```suggestion
-- TODO: Should we create a macro?
```

**Comment by semorrison** [comment]
I have a local build of Mathlib I'm happy with.

File: src/Init/Data/UInt/Lemmas.lean [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+theorem zero_def : (0 : $typeName) = ⟨0⟩ := rfl
+theorem one_def : (1 : $typeName) = ⟨1⟩ := rfl
+theorem sub_def (a b : $typeName) : a - b = ⟨a.val - b.val⟩ := rfl
+theorem mul_def (a b : $typeName) : a * b = ⟨a.val * b.val⟩ := rfl
+theorem mod_def (a b : $typeName) : a % b = ⟨a.val % b.val⟩ := rfl
+theorem add_def (a b : $typeName) : a + b = ⟨a.val + b.val⟩ := rfl
+
+@[simp] theorem mk_val_eq : ∀ (a : $typeName), mk a.val = a
+| ⟨_, _⟩ => rfl
```
Comment (david-christiansen) [inline src/Init/Data/UInt/Lemmas.lean:26]:
```suggestion
  | ⟨_, _⟩ => rfl
```
For consistency with typical style

File: src/Init/Data/UInt/Lemmas.lean [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+theorem val_eq_of_lt {a : Nat} : a < size → ((ofNat a).val : Nat) = a :=
+        Nat.mod_eq_of_lt
```
Comment (david-christiansen) [inline src/Init/Data/UInt/Lemmas.lean:28]:
```suggestion
  Nat.mod_eq_of_lt
```

**Pull request merged**